core/WASM_core.md

```diff
@@ -0,0 +1,352 @@
+  network calls. We'll test this, though. Even if we intend to use native gRPC clients, being able
+  to make network calls in Core is still pretty desirable for things like auto-updates, test server
+  downloads, etc.
+* Calling back into lang gives us opportunities for more user customization, which can be nice.
+  Especially when it comes to the metrics and tracing stuff. gRPC interceptors are also a potential
+  plus.
+
+My inclination for now is to at least *investigate* how much we can get out of WASI. Particularly
+with respect to running from Go.
+
+Unfortunately, my assumption is that we will need to callback into Lang for all the usecases
+mentioned at the top of the doc. The only sys interface things it appears we'll be able to reliably
+use are clocks (which is good, that does come up) and possibly file I/O, which currently does not
+happen in Core (and probably never should) except for in the test server usecase.
+
+
+### Core API changes
+Operating under the assumption that all the major syscalls will need to be handled through callbacks
+into lang. In WASM, calling functions in the host is supported via `import`ing those host functions.
+Ultimately these functions may be bundled together into a trait or something on the Core side, but
+this doc will focus on the individual functions.
+
+Here are the functions I think we'll need to add to the Core<-->Lang interface, beyond the existing
+ones that lang already calls. They're defined as they would appear in Rust, after being imported.
+This is definitely a rough draft, as it will be influenced by runtime selection and whatever
+practical problems we run into along the way.
+
+```rust
+// In reality, everything in these functions is going to need to be serialized and shuttled back
+// and forth as bytes. That makes for uninteresting signatures, though, so instead I show what
+// those signatures would look like if we were using deserialized types directly. The exact details
+// of how that serialization happens are intentionally not included here.
+//
+// WASM has an `ExternRef` type which is a pointer to a host-owned opaque object.
+
+// For all the types which wrap an `ExternRef`, the implementations on those types which take
+// `self` will be implemented as calls to a host-defined function which looks like:
+fn func_name(receiver: ExternRef, arg: ArgType) -> ResultType;
+// There may be a more clever way to manage some of this by using `Table` and `Func` types along with
+// `ExternRef` but I need to play around more to determine that.
+
+// -------------- Grpc -----------------
+
+// Would be passed into core along with client config options when creating instantiating a client.
+// Core can then augment the client with retry behavior, and the client can then be used when
+// instantiating Core workers.
+struct GrpcClientRef {
+    /// Points to the host's gRPC client
+    host_ref: ExternRef,
+}
+
+struct GrpcRequest {
+    uri: String,
```

Comment:
What is URI here? I think this should be two strings, "service" and "method" in gRPC parlance. Most users of this will have to have a big switch statement for this to make typed calls from their lang (though many gRPC impls have a https://pkg.go.dev/google.golang.org/grpc#ClientConn.Invoke like dynamic call, you often still need to know the response type). And since this would be one big switch, if you wanted to instead make gRPC request service+method one enum of needed methods, that may work better than string.

Reply:
I wasn't sure if we could use raw bytes for the inputs and outputs, seems like you're saying we can't.

Reply:
We absolutely can use raw bytes to/from core (and should, we only want a single gRPC invoke call). But most lang sides probably will have a big switch statement on the method name to convert to protos they are aware of. I think most (all?) gRPC client libraries intentionally hide the byte-level communication so they have complete control over the protocol.

Reply:
I think this big switch is fine (some langs can employ code gen or reflection if they must). Benefit of using the gRPC client from the outside with its expected methods and req/resp types is that you get all the other gRPC client benefits such as interceptors and advanced dial/call options.